nova/objects/instance.py

```diff
@@ -1426,11 +1428,29 @@ def fill_faults(self):
         :returns: A list of instance uuids for which faults were found.
         """
         uuids = [inst.uuid for inst in self]
-        faults = objects.InstanceFaultList.get_latest_by_instance_uuids(
+        inst_mappings = objects.InstanceMappingList.get_by_instance_uuids(
             self._context, uuids)
+
+        def _group_inst_mappings_by_cell_id(instance_mappings):
+            def _im_cell_id(instance_mapping):
+                cm = instance_mapping.cell_mapping
+                cell_id = getattr(cm, 'id', None) if cm else None
+                return cell_id
+
+            instance_mappings = sorted(instance_mappings, key=_im_cell_id)
+            return itertools.groupby(instance_mappings, key=_im_cell_id)
+
         faults_by_uuid = {}
-        for fault in faults:
-            faults_by_uuid[fault.instance_uuid] = fault
+        for cell_id, ims in _group_inst_mappings_by_cell_id(inst_mappings):
+            cm, inst_uuids = None, []
+            for im in ims:
+                inst_uuids.append(im.instance_uuid)
+                cm = im.cell_mapping
```

Comment:
Do I understand this correctly: `cell_id` can be `None` and `im.cell_mapping` can be `None`, too? Should we have an `if cell_id is None: continue` in the for-loop?

Reply:
Yes. PR updated.

Reply:
You're missing a test for this. You updated the tests, so the cell-mapping is there, but nothing failed before, so we're missing a test.